bodhi/server/util.py

```diff
@@ -249,62 +248,110 @@ def sanity_check_repodata(myurl, repo_type):
-        if len(comps.groups) < 1:
-            raise RepodataException('Comps file empty')
-
-    # Test updateinfo
-    ret = subprocess.call(['zgrep', '<id/>', repo_info['updateinfo']])
-    if not ret:
-        raise RepodataException('updateinfo.xml.gz contains empty ID tags')
+            reqparts.append('modules')
+        missing = []
+        for part in reqparts:
+            if part not in repo_info:
+                missing.append(part)
+        if missing:
+            raise RepodataException(f'Required parts not in repomd.xml: {", ".join(missing)}')
+
+        # Only yum repos have comps
+        if repo_type == 'yum':
+            # Test comps
+            comps = libcomps.Comps()
+            try:
+                ret = comps.fromxml_f(repo_info['group'])
+            except Exception:
+                raise RepodataException('Comps file unable to be parsed')
+            if len(comps.groups) < 1:
+                raise RepodataException('Comps file empty')
+
+        # Test updateinfo
+        ret = subprocess.call(['zgrep', '<id/>', repo_info['updateinfo']])
+        if not ret:
+            raise RepodataException('updateinfo.xml.gz contains empty ID tags')
+
+        # Now call out to DNF to check if the repo is usable
+        # "tests" is a list of tuples with (dnf args, expected output) to run.
+        # For every test, DNF is run with the arguments, and if the expected output is not found,
+        #  an error is raised.
+        tests = []
+
+        if repo_type in ('yum', 'source'):
+            tests.append((['list', 'available'], 'testrepo'))
+        elif repo_type == 'module':
+            tests.append((['module', 'list'], '.*'))
+
+        for test in tests:
+            dnfargs, expout = test
+
+            # Make sure every DNF test runs in a new temp dir
+            testdir = tempfile.mkdtemp(dir=tmpdir)
+            output = sanity_check_repodata_dnf(testdir, myurl, *dnfargs)
+            if (expout == ".*" and len(output.strip()) != 0) or (expout in output):
+                continue
+            else:
+                raise RepodataException(
+                    f"DNF did not return expected output when running test!"
+                    + f" Test: {dnfargs}, expected: {expout}, output: {output}")
+
+
+def sanity_check_repodata_dnf(tempdir, myurl, *dnf_args):
+    """
+    Call DNF to try to parse and sanity check the repository.
+
+    Because DNF does not provide any usable API they won't break in a year time, we just shell out
+    to run some repository actions on it.
+
+    Args:
+        tempdir (basestring): Temporary directory that will be removed at the end.
+        myurl (basestring): A path to a repodata directory.
+        dnf_args (list): A list of arguments after DNF is set up to use the repo under test.
+            Example would be ["search", "kernel"] or ["module", "list"].
+    Raises:
+        Exception: If the repodata is not valid or does not exist.
```

Comment:
Actually, [function](https://docs.python.org/3/library/subprocess.html#subprocess.check_output) is returning [CalledProcessError](https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError)

Reply:
For now that's the only option, yes. But I have plans to add more checks, which might return other exceptions.
And the current docblock is technically correct:

```
>>> issubclass(subprocess.SubprocessError, Exception)
True
```